src/core/models.py

```diff
@@ -5,51 +5,87 @@
 """Collection of state objects for the Kafka relations, apps and units."""
 
 import logging
-from typing import MutableMapping
+from typing import Literal, MutableMapping
 
+from charms.data_platform_libs.v0.data_interfaces import Data, DataPeerData, DataPeerUnitData
 from charms.zookeeper.v0.client import QuorumLeaderNotFoundError, ZooKeeperManager
-from kazoo.exceptions import AuthFailedError, NoNodeError
+from kazoo.client import AuthFailedError, NoNodeError
 from ops.model import Application, Relation, Unit
-from tenacity import retry, retry_if_not_result, stop_after_attempt, wait_fixed
+from tenacity import retry
+from tenacity.retry import retry_if_not_result
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+from typing_extensions import override
 
-from literals import INTERNAL_USERS, Substrate
+from literals import INTERNAL_USERS, SECRETS_APP
 
 logger = logging.getLogger(__name__)
 
+SUBSTRATES = Literal["vm", "k8s"]
```

Comment:
**nitpick** (non-blocking): literals and other type hints usually use camel case rather than an all-caps form like a constant.